infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java

```diff
@@ -203,109 +172,117 @@ private CorfuRuntime getNewCorfuRuntime() {
         return runtime;
     }
 
-    private boolean isBootstrapped(CorfuMsg msg) {
+    private boolean isBootstrapped(RequestMsg req) {
         if (serverContext.getManagementLayout() == null) {
-            log.warn("Received message but not bootstrapped! Message={}", msg);
+            log.warn("Received message but not bootstrapped! Message={}", req.getHeader());
             return false;
         }
+
         return true;
     }
 
     /**
      * Forward an orchestrator request to the orchestrator service.
      *
-     * @param msg corfu message containing ORCHESTRATOR_REQUEST
-     * @param ctx netty ChannelHandlerContext
-     * @param r   server router
+     * @param req  a message containing an ORCHESTRATOR request
+     * @param ctx  the netty ChannelHandlerContext
+     * @param r    the server router
      */
-    @ServerHandler(type = CorfuMsgType.ORCHESTRATOR_REQUEST)
-    public synchronized void handleOrchestratorMsg(@Nonnull CorfuPayloadMsg<OrchestratorMsg> msg,
-                                                   @Nonnull ChannelHandlerContext ctx,
-                                                   @Nonnull IServerRouter r) {
-        log.debug("Received an orchestrator message {}", msg);
-        orchestrator.handle(msg, ctx, r);
+    @RequestHandler(type = PayloadCase.ORCHESTRATOR_REQUEST)
+    public synchronized void handleOrchestratorMsg(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        log.debug("handleOrchestratorMsg: message:{}", req.getPayload().getOrchestratorRequest());
+        orchestrator.handle(req, ctx, r);
     }
 
     /**
      * Bootstraps the management server.
-     * The msg contains the layout to be bootstrapped.
+     * The msg contains the Layout to be bootstrapped with.
      *
-     * @param msg corfu message containing MANAGEMENT_BOOTSTRAP_REQUEST
-     * @param ctx netty ChannelHandlerContext
-     * @param r   server router
+     * @param req  a message containing a BOOTSTRAP_MANAGEMENT request
+     * @param ctx  the netty ChannelHandlerContext
+     * @param r    the server router
      */
-    @ServerHandler(type = CorfuMsgType.MANAGEMENT_BOOTSTRAP_REQUEST)
-    public synchronized void handleManagementBootstrap(CorfuPayloadMsg<Layout> msg,
-                                                       ChannelHandlerContext ctx, IServerRouter r) {
-        if (serverContext.getManagementLayout() != null) {
-            // We are already bootstrapped, bootstrap again is not allowed.
-            log.warn("handleManagementBootstrap: Got a request to bootstrap a server " +
-                    "with {} which is already bootstrapped, rejecting!", msg.getPayload());
-            r.sendResponse(ctx, msg, new CorfuMsg(CorfuMsgType.MANAGEMENT_ALREADY_BOOTSTRAP_ERROR));
+    @RequestHandler(type = PayloadCase.BOOTSTRAP_MANAGEMENT_REQUEST)
+    public synchronized void handleBootstrapManagement(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
```

Comment:
Is there any specific criteria that some handlers are set to `synchronized` but some are not? It looks to me that the operations upon shared resources have been synchronized in a lower level, (i.e. methods in servercontext, some concurrenthashmap, etc), and all the synchronized signature could be removed?

Reply:
I am not sure if it's correct, but this is my reasoning. It does not always suffice to consider that the shared resources are themselves synchronized, but one also needs to look at the interaction between them. Consider the following scenario. The management server receives a `BOOTSTRAP_MANAGEMENT` request with `Layout` A. Since it isn't bootstrapped yet, the call to `getManagementLayout` will return `null` and the thread will enter the `else` clause. Now, before the thread has a chance to call `setManagementLayout`, it goes to sleep. Now the ManagementServer receives another `BOOTSTRAP_MANAGEMENT` request with `Layout` B. Since the first request did not complete, this thread will also get `null` from the call to `getManagementLayout` and enter the `else` clause. Now we are in the state where there are two bootstrap requests with different layouts calling `setManagementLayout`. If my understanding it correct, then this is problematic and leaving the `synchronized` keyword here will prevent such an issue.

Reply:
Good point, this was actually a production bug at point.

Sometimes you need a bigger lock to protect against interleaving with other "synchronized APIs". For example, think about the local data store where put/get are individually synchronized, but there are some code paths that do something like:

```java

val = datastore.get(key)
// some computation on val
val2 = someFunction(val)
datastore.put(key, val2)
````

This is a read-modify-write cycle where interleaving with another thread between the get and put can cause some inconsistency. Notice that this consistency can happen despite each API being thread-safe individually.

Although this style of synchronization can be safe when implemented carefully, its usually fragile. Ideally, the datastore should expose an interface for some notion of a transaction, where multiple operations can be executed as a single atomic unit.